// @azure/cosmos/dist-esm/src/request/hybridSearchQueryResult.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
const FieldNames = {
    Rid: "_rid",
    Payload: "payload",
    ComponentScores: "componentScores",
};
export class HybridSearchQueryResult {
    constructor(rid, componentScores, data) {
        this.rid = rid;
        this.componentScores = componentScores;
        this.data = data;
    }
    static create(document) {
        const rid = document[FieldNames.Rid];
        if (!rid) {
            throw new Error(`${FieldNames.Rid} must exist.`);
        }
        const outerPayload = document[FieldNames.Payload];
        if (!outerPayload || typeof outerPayload !== "object") {
            throw new Error(`${FieldNames.Payload} must exist.`);
        }
        const innerPayload = outerPayload[FieldNames.Payload];
        if (!innerPayload || typeof innerPayload !== "object") {
            throw new Error(`${FieldNames.Payload} must exist nested within the outer payload field.`);
        }
        const componentScores = outerPayload[FieldNames.ComponentScores];
        if (!Array.isArray(componentScores)) {
            throw new Error(`${FieldNames.ComponentScores} must exist.`);
        }
        return new HybridSearchQueryResult(rid, componentScores, innerPayload);
    }
}
